src/UserSpaceInstrumentation/include/UserSpaceInstrumentation/RegisterState.h

```diff
@@ -0,0 +1,240 @@
+  uint32_t edx;
+  uint32_t esi;
+  uint32_t edi;
+  uint32_t ebp;
+  uint32_t eax;
+  uint32_t xds;
+  uint32_t xes;
+  uint32_t xfs;
+  uint32_t xgs;
+  uint32_t orig_eax;
+  uint32_t eip;
+  uint32_t xcs;
+  uint32_t eflags;
+  uint32_t esp;
+  uint32_t xss;
+};
+
+struct GeneralPurposeRegisters64 {
+  uint64_t r15;
+  uint64_t r14;
+  uint64_t r13;
+  uint64_t r12;
+  uint64_t rbp;
+  uint64_t rbx;
+  uint64_t r11;
+  uint64_t r10;
+  uint64_t r9;
+  uint64_t r8;
+  uint64_t rax;
+  uint64_t rcx;
+  uint64_t rdx;
+  uint64_t rsi;
+  uint64_t rdi;
+  uint64_t orig_rax;
+  uint64_t rip;
+  uint64_t cs;
+  uint64_t rflags;
+  uint64_t rsp;
+  uint64_t ss;
+  uint64_t fs_base;
+  uint64_t gs_base;
+  uint64_t ds;
+  uint64_t es;
+  uint64_t fs;
+  uint64_t gs;
+};
+
+union GeneralPurposeRegisters {
+  GeneralPurposeRegisters32 x86_32;
+  GeneralPurposeRegisters64 x86_64;
+};
+
+struct MmsAs80BitFloat {
+  uint64_t mantissa;
+  uint16_t sign_exp;
+};
+static_assert(sizeof(MmsAs80BitFloat) == 10, "MmsAsFloat is not 10 bytes of size");
+
+struct MmsRegister {
+  union {
+    std::array<uint8_t, 10> bytes;
+    MmsAs80BitFloat as_float;
+  };
+  std::array<uint8_t, 6> reserved;
+};
+static_assert(sizeof(MmsRegister) == 16, "MmsRegister is not 16 bytes of size");
+
+struct XmmRegister {
+  uint8_t bytes[16];
+};
+
+// Legacy Region of the XSave area stores the Fpu, Mmx, Sse state of the Cpu.
+// Compare "Intel 64 and IA-32 Architectures Software Developer’s Manual, Volume 1" section 13.4.1.
+struct FxSave {
+  uint16_t fcw;
+  uint16_t fsw;
+  uint16_t ftw;
+  uint16_t fop;
+  uint64_t fip;
+  uint64_t fdp;
+  uint32_t mxcsr;
+  uint32_t mxcsr_mask;
+  std::array<MmsRegister, 8> stmm;
+  std::array<XmmRegister, 16> xmm;
+  std::array<uint8_t, 48> padding1;
+  uint64_t xcr0;
+  std::array<uint8_t, 40> padding2;
+};
+static_assert(sizeof(FxSave) == 512, "FxSave is not 512 bytes of size");
+
+// XSave header contains information about what is present in the extended region of an XSAVE area.
+// Compare "Intel 64 and IA-32 Architectures Software Developer’s Manual, Volume 1" section 13.4.2.
+// and 13.4.3..
+struct XSaveHeader {
+  enum class StateComponents : uint64_t {
+    kX87 = 1,
+    kSse = 2,
+    kAvx = 4,
+    kBndRegs = 8,
+    kBndCsr = 16,
+    kOpMask = 32,
+    kZmmHi256 = 64,
+    kHi16Zmm = 128,
+    kPt = 256,
+    kPkru = 512,
+  };
+  uint64_t xstate_bv;
+  uint64_t xcomp_bv;
+  std::array<uint64_t, 6> reserved;
+};
+static_assert(sizeof(XSaveHeader) == 64, "XSaveHeader layout incorrect");
+
+// The upper 128 bit of a single YMMx register.
+struct YmmHiRegister {
+  std::array<uint8_t, 16> bytes;
+};
+
+// The upper 128 bit of the YMM0..15 registers. The lower bits are shared with XMM registers above.
+struct YmmHi {
+  std::array<YmmHiRegister, 16> ymm;
+};
+PACKED_END
+
+// Backup, modify and restore register state of a halted thread.
+//
+// Requires the XSave extension (added around 2008) to be supported by the Cpu otherwise
```

Comment:
I think that the piece missing in all the comments here is a simple definition of the "XSave area". One gets the idea but it would be clearer to have an explanation right away.

Reply:
Added some line here. done.